### ISSUE TITLE
Leibniz equality confuses "Add Ring"

### ISSUE THREAD

**Issue by Zimmi48** [Zimmi48/bugzilla-test#1448]
Note: the issue was created automatically with bugzilla2github tool

# Bugzilla Bug ID: 1448
Date: 2007-03-19 14:02:37 +0100
From: Roland.Zumkeller@polytechnique.fr
To:   
CC:   emakarov@gmail.com

Last updated: 2007-10-30 18:09:57 +0100

## Bugzilla Comment ID: 112
Date: 2007-03-19 14:02:37 +0100
From: Roland.Zumkeller@polytechnique.fr

Full_Name: Roland Zumkeller
Version: 8.1
OS: Mac OS X
Submission from: sphinx.lix.polytechnique.fr (129.104.11.1)

Require Import Relations.
Require Import Ring_theory.

Variable R : Type.
Variable Rone Rzero : R.
Variable Rplus Rmult Rminus : R -> R -> R.
Variable Rneg : R -> R.

Lemma my_ring_theory : @ring_theory R Rzero Rone Rplus Rmult Rminus Rneg (@eq
R).
Admitted.

Add Ring my_ring : my_ring_theory (abstract).

Anomaly: Ring: cannot find hypo in modules Coq.setoid_ring.Ring_theory,
 Coq.setoid_ring.Ring_polynom, Coq.setoid_ring.Ring_tac,
 Coq.setoid_ring.InitialRing, Coq.setoid_ring.Field_tac,
 Coq.setoid_ring.Field_theory. Please report.

Furthermore, declaring a relation Req on R before "Add Ring" yields the
following:

Variable Req : R -> R -> Prop.

Hypothesis Req_refl : reflexive _ Req.
Hypothesis Req_sym : symmetric _ Req.
Hypothesis Req_trans : transitive _ Req.

Add Relation R Req
  reflexivity proved by Req_refl
  symmetry proved by Req_sym
  transitivity proved by Req_trans
  as Req_rel.

Add Ring my_ring : my_ring_theory (abstract).

User error: ring addition should be declared as a morphism

This is unexpected, since my_ring_theory refers to Leibniz equality.

## Bugzilla Comment ID: 114
Date: 2007-03-20 08:32:09 +0100
From: Roland.Zumkeller@polytechnique.fr

Benjamin Gregoire <Benjamin.Gregoire@sophia.inria.fr>
Thank Roland, this is a but.
   Require Import InitialRing.
   We should be done before adding a new ring, we should insert a sanity 
check.
 Benjamin

Roland.Zumkeller@polytechnique.fr wrote:

>Full_Name: Roland Zumkeller
>Version: 8.1
>OS: Mac OS X
>Submission from: sphinx.lix.polytechnique.fr (129.104.11.1)
>
>
>Require Import Relations.
>Require Import Ring_theory.
>
>Variable R : Type.
>Variable Rone Rzero : R.
>Variable Rplus Rmult Rminus : R -> R -> R.
>Variable Rneg : R -> R.
>
>Lemma my_ring_theory : @ring_theory R Rzero Rone Rplus Rmult Rminus Rneg (@eq
>R).
>Admitted.
>
>Add Ring my_ring : my_ring_theory (abstract).
>
>Anomaly: Ring: cannot find hypo in modules Coq.setoid_ring.Ring_theory,
> Coq.setoid_ring.Ring_polynom, Coq.setoid_ring.Ring_tac,
> Coq.setoid_ring.InitialRing, Coq.setoid_ring.Field_tac,
> Coq.setoid_ring.Field_theory. Please report.
>
>
>Furthermore, declaring a relation Req on R before "Add Ring" yields the
>following:
>
>Variable Req : R -> R -> Prop.
>
>Hypothesis Req_refl : reflexive _ Req.
>Hypothesis Req_sym : symmetric _ Req.
>Hypothesis Req_trans : transitive _ Req.
>
>Add Relation R Req
>  reflexivity proved by Req_refl
>  symmetry proved by Req_sym
>  transitivity proved by Req_trans
>  as Req_rel.
>
>Add Ring my_ring : my_ring_theory (abstract).
>
>User error: ring addition should be declared as a morphism
>
>This is unexpected, since my_ring_theory refers to Leibniz equality.
>
>
>  
>

## Bugzilla Comment ID: 115
Date: 2007-03-20 08:40:11 +0100
From: Roland.Zumkeller@polytechnique.fr

"Roland Zumkeller" <roland.zumkeller@gmail.com>
Thanks for the quick reply.

This makes the first case work, but in the second case it still asks
for a morphism:

Require Import Relations.
Require Import Ring_theory.
Require Import InitialRing.

Variable R : Type.
Variable Rone Rzero : R.
Variable Rplus Rmult Rminus : R -> R -> R.
Variable Rneg : R -> R.

Lemma my_ring_theory : @ring_theory R Rzero Rone Rplus Rmult Rminus Rneg (@eq
R).
Admitted.

Variable Req : R -> R -> Prop.

Hypothesis Req_refl : reflexive _ Req.
Hypothesis Req_sym : symmetric _ Req.
Hypothesis Req_trans : transitive _ Req.

Add Relation R Req
  reflexivity proved by Req_refl
  symmetry proved by Req_sym
  transitivity proved by Req_trans
  as Req_rel.

Add Ring my_ring : my_ring_theory (abstract).

User error: ring addition should be declared as a morphism

## Bugzilla Comment ID: 113
Date: 2007-07-04 17:02:35 +0200
From: Roland.Zumkeller@polytechnique.fr

BB ?

## Bugzilla Comment ID: 1699
Date: 2007-08-06 14:06:51 +0200
From: @_barras

computing carrier from the relation type was not right (did not expand constants)

## Bugzilla Comment ID: 1887
Date: 2007-10-30 13:30:56 +0100
From: Jean-Marc Notin &lt;<notin@lix.polytechnique.fr>&gt;

The latest stable version of Coq (8.1pl2) includes the bugfix => this bug is now closed.

## Bugzilla Comment ID: 1894
Date: 2007-10-30 18:09:57 +0100
From: Jean-Marc Notin &lt;<notin@lix.polytechnique.fr>&gt;

*** Bug #1674 has been marked as a duplicate of this bug. ***
